soccer/gameplay/tactics/advance_zone_midfielder.py

```diff
@@ -0,0 +1,186 @@
+            "When a kick play begins")
+        self.add_transition(
+            AdvanceZoneMidfielder.State.hold,
+            AdvanceZoneMidfielder.State.pass_set, lambda: not self.kick,
+            "When not in a kick play")
+
+    def execute_pass_set(self):
+        # gets the best position to travel to for ball reception
+        points = [robocup.Point(0, 0), robocup.Point(0, 0)]
+        best_point = self.passing_point
+
+        # sets the second point
+        alt_point, value2 = evaluation.passing_positioning.eval_best_receive_point(
+                self.passing_point,
+                main.our_robots(), AdvanceZoneMidfielder.FIELD_POS_WEIGHTS,
+                AdvanceZoneMidfielder.NELDER_MEAD_ARGS,
+                AdvanceZoneMidfielder.PASSING_WEIGHTS)
+
+        # check for futile position i.e the alternate position is in the way of a shot from best position
+        if self.in_shot_triangle(best_point, alt_point):
+            alt_point = self.remove_obstruction(best_point, alt_point)
+
+        points = [best_point, alt_point]
+        # moves the robots and assigns information
+        for i in range(len(points)):
+            if (self.moves[i] is None):
+                self.moves[i] = skills.move.Move(points[i])
+                self.add_subbehavior(
+                    self.moves[i],
+                    self.names[i],
+                    required=False,
+                    priority=self.priorities[i])
+            else:
+                self.moves[i].pos = points[i]
+    #defaults to a defensive position on kicks
+    def execute_hold(self):
+        # old methodology for simple zone midfielder (check comments on that method)
+        y_temp_hold = 0.8 * self.passing_point.y
+
+        x_temp_hold = constants.Field.Width / 3
+
+        self.hold_point = [
+            robocup.Point(x_temp_hold, y_temp_hold),
+            robocup.Point(-x_temp_hold, y_temp_hold)
+        ]
+
+        for i in range(len(self.hold_point)):
+            if (self.moves[i] is None):
+                self.moves[i] = skills.move.Move(self.hold_point[i])
+                self.add_subbehavior(
+                    self.moves[i], self.names[i], required=False, priority=1)
+            else:
+                self.moves[i].pos = self.hold_point[i]
+
+    def on_exit_pass_set(self):
+        self.remove_all_subbehaviors()
+
+    def on_exit_hold(self):
+        self.remove_all_subbehaviors()
+
+    # this method determines if the 2nd midfielders kicking point is actually in the way of the potential shot
+    def in_shot_triangle(self, best_point, alt_point):
+        # get the two points of the enemies goal
+        goalSegment = constants.Field.TheirGoalSegment
+        goalCenter = goalSegment.center()
+        length = goalSegment.length()
+        right_post = robocup.Point(goalCenter.x + length / 2, goalCenter.y)
+        left_post = robocup.Point(goalCenter.x - length / 2, goalCenter.y)
```

Comment:
You can do goalSegment.pt[0] and goalSegment.pt[1] here instead of the goalCenter + length/2